### ISSUE TITLE
Wrapped methods are not properly resolved until after the class definition

### ISSUE THREAD

**Issue by VincentVanlaer** [python/mypy#16397]
# Context & setup

Consider the following simple protocol:

```python
class Proto(Protocol):
    @property
    def f(self) -> int:
        ...
```

One can implement this protocol wrongly as follows:

```python
T = TypeVar("T")
P = ParamSpec("P")


def wrapper(f: Callable[P, T]) -> Callable[P, str]:
    def wrapped(*args: P.args, **kwargs: P.kwargs) -> str:
        return 0

    return wrapped

class Impl:
    @wrapper  # converts return type from `bool` to `str`, while the protocol expects `int`
    def f(self) -> bool:
        return False

# error: Incompatible return value type (got "Impl", expected "Proto")  [return-value]
# note: Following member(s) of "Impl" have conflicts:
# note:     Expected:
# note:         def f(self) -> int
# note:     Got:
# note:         def f(self) -> str
def b() -> Proto:
    return Impl()

```

# The bug (false negative)

When this same function `b` is placed before `class Impl`, [mypy does not report an error](https://mypy-play.net/?mypy=master&python=3.12&gist=28a8e417ff96b3f16b716f22a5bfb602&flags=strict%2Cdisallow-any-decorated%2Cdisallow-any-expr%2Cdisallow-any-explicit%2Cdisallow-any-generics%2Cdisallow-any-unimported). Adding `reveal_type(Impl.f)` before and after the class definition hints at what is going on:

```
main.py:21: error: Expression has type "Any"  [misc]
main.py:21: error: Cannot determine type of "f"  [has-type]
main.py:21: note: Revealed type is "Any"
main.py:21: error: Name "Impl" is used before definition  [used-before-def]

main.py:28: note: Revealed type is "def (self: __main__.Impl) -> builtins.str"
```

If `f` is not decorated but explicitly returns `str`, one gets

```
main.py:21: note: Revealed type is "def (self: __main__.Impl) -> builtins.str"
main.py:21: error: Name "Impl" is used before definition  [used-before-def]

main.py:27: note: Revealed type is "def (self: __main__.Impl) -> builtins.str"
```
as expected. This indicates that the `wrapped` decorator is only properly resolved for code after the class definition.

# False positive variant

If `f` is decorated with `@property`, this can also produce false positives. This is how I encountered the issue in the first place. [Playground.](https://mypy-play.net/?mypy=master&python=3.12&gist=d0f9e30f418117f6acec20af3caa57fc&flags=strict%2Cdisallow-any-decorated%2Cdisallow-any-expr%2Cdisallow-any-explicit%2Cdisallow-any-generics%2Cdisallow-any-unimported)

# Additional info

Python: `3.11`
Mypy: 371219347a6d17e16924bbabf3e693c6874e7138
Flags: `--strict --disallow-any-*`
